development/extensions/tutorial_modules.rst

```diff
@@ -0,0 +1,319 @@
+            );
+        }
+    }
+
+The ``main_info`` class has only one method ``module()`` which returns an array
+with the following information:
+
+.. csv-table::
+    :header: "Key", "Content"
+    :delim: |
+
+    ``filename`` | "Fully name-spaced path to the `Module class`_, starting with a leading slash."
+    ``title`` | "Language key for the module-title that is displayed in the ACP module
+    management section."
+    ``modes`` | "Contains an array of modes for the module. See the next table
+    for more details."
+
+Note that we have defined one mode named ``settings`` with the following information:
+
+.. csv-table::
+    :header: "Key", "Content"
+    :delim: |
+
+    ``title`` | "Language key of the mode-title that is displayed in the ACP module
+    management section."
+    ``auth`` | "An authorisation key used to control access to the mode.
+    In this case we require the extension ``acme/demo`` to be enabled and
+    that the user have the ``a_board`` permission."
+    ``cat`` | "References the parent category the mode belongs to. Should
+    typically contain the language key of the parent category."
+
+Module class
+------------
+
+The module class contains the actual code for the module. This
+file will also be stored in the ``acp/`` folder of the extension.
+For this tutorial, we will use ``ext/acme/demo/acp/main_module.php``:
+
+.. code-block:: php
+
+    <?php
+
+    namespace acme\demo\acp;
+
+    class main_module
+    {
+        public $u_action;
+        public $tpl_name;
+        public $page_title;
+
+        public function main($id, $mode)
+        {
+            global $user, $template, $request, $config;
+
+            $this->tpl_name = 'acp_demo_body';
+            $this->page_title = $user->lang('ACP_DEMO_TITLE');
+
+            add_form_key('acme/demo');
+
+            if ($request->is_set_post('submit'))
+            {
+                if (!check_form_key('acme/demo'))
+                {
+                     trigger_error('FORM_INVALID');
+                }
+
+                $config->set('acme_demo_goodbye', $request->variable('acme_demo_goodbye', 0));
+                trigger_error($user->lang('ACP_DEMO_SETTING_SAVED') . adm_back_link($this->u_action));
+            }
+
+            $template->assign_vars(array(
+                'ACME_DEMO_GOODBYE' => $config['acme_demo_goodbye'],
+                'U_ACTION'          => $this->u_action,
+            ));
+        }
+    }
+
+The module itself must contain a ``main($id, $mode)`` method,
+which takes the ``id`` of the module in the database and the ``mode``
+being called as arguments.
+
+In the code of the ``main`` method, we first set the name of the
+template file that is used to render the module, and the page
+title is assigned a language key.
+
+.. _form key:
+
+To strengthen the form against security vulnerabilities, we
+use a form key check to verify that the form being submitting
+is valid. This is done by calling ``add_form_key('acme/demo')``
```

Comment:
Add a warning that the parameter for add_form_key should be unique for every form.